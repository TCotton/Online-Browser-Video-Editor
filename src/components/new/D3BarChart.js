const D3BarChart = {};

D3BarChart.create = (el, data, configuration) => {

}

D3BarChart.update = (el, data, configuration, chart) => {

}

D3BarChart.destroy = () => {

}

export default D3BarChart;