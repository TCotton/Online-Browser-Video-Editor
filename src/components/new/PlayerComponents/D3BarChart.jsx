import React, {useEffect, useRef} from "react";
import * as d3 from "d3";

const D3BarChart = props => {
    const {randomData, width, height, padding, colour} = props;

    const ref = useRef()

    useEffect(() => {
        const svgElement = d3.select(ref.current);
        const result = [{h: randomData, c: colour}];

        const x = d3.scaleBand().range([0, width]).domain(result.map(d => d.c)).padding(padding);
        const y = d3.scaleLinear().range([height, 0]).domain([1, 255]);

        const svg = svgElement.selectAll('.rects').selectAll("rect").data(result);

        svg.join(
            enter => {
                enter.append('rect')
                    .attr('width', x.bandwidth)
                    .attr("height", d => {
                        if (d && d.h) {
                            return height - y(d.h);
                        }
                        return 0;
                    })
                    .attr('y', (d) => {
                        if (d && d.h) {
                            return y(d.h);
                        }
                        return 0;
                    })
                    .attr('fill', d => {
                        if (d && d.c) {
                            return d.c;
                        }
                        return null;
                    })
                    .attr('class', 'rect');
            },
            update => {
                update
                    .attr('width', x.bandwidth)
                    .attr('y', (d) => {
                        if (d && d.h) {
                            return y(d.h);
                        }
                        return 0;
                    })
                    .attr("height", d => {
                        if (d && d.h) {
                            return height - y(d.h);
                        }
                        return 0;
                    })
            },
            destroy => {
                destroy.exit().remove();
            }
        )

        const svgTicks = svgElement.selectAll('.left-ticks').selectAll("text").data(result);

        const axlft = d3.axisLeft(y).ticks(10, 's');

        svgTicks.join(
            enter => {
                enter
                    .append('text').text("value").attr('font-size', 14)
            }
        ).call(axlft);


    }, [randomData])

    return (
        <svg
            height={`100%`}
            width={`100%`}
            viewBox={`0 0 ${width} ${height}`}
            preserveAspectRatio="none"
            ref={ref}
        >
            <g className="rects" transform="translate(0,0)"/>
            <g className="left-ticks" transform="translate(20,20)"/>
        </svg>
    );
};

export default D3BarChart;
