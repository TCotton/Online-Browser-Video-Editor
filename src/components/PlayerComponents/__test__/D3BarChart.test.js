import React from 'react';
import {render} from '@testing-library/react';
import D3BarChart from "../D3BarChart";

describe('D3BarChart', () => {

    const props = {
        setClass: jest.fn(),
        colour: '#000',
        padding: 10,
        height: 10,
        width: 10,
        data: 123
    }

    it('should render successfully', () => {
        const {container} = render(<D3BarChart {...props}/>);
        expect(container.querySelector("[data-testid=\"svg\"]")).not.toBeNull();
    });

    it('should match snapshot', () => {
        const {baseElement} = render(<D3BarChart {...props}/>);
        expect(baseElement).toMatchSnapshot();
    });

    it('should call function setClass if data prop is more than 0',  () => {
        const mockFnc = jest.fn();
        props.setClass = mockFnc;
        render(<D3BarChart {...props}/>);
        expect(mockFnc.mock.calls.length).toBe(1);
    });

    it('should not call function is data prop is 0', () => {
        const mockFnc = jest.fn();
        props.setClass = mockFnc;
        props.data = 0
        render(<D3BarChart {...props}/>);
        expect(mockFnc.mock.calls.length).toBe(0);
    });

});