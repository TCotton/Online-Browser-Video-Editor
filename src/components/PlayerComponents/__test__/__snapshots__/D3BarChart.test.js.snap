// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D3BarChart should match snapshot 1`] = `
<body>
  <div>
    <svg
      data-testid="svg"
      height="100%"
      preserveAspectRatio="none"
      viewBox="0 0 10 10"
      width="100%"
    >
      <g
        class="rects"
        data-testid="rects"
        transform="translate(0,0)"
      >
        <rect
          class="rect"
          fill="#000"
          height="4.803149606299213"
          width="0"
          y="5.196850393700787"
        />
      </g>
    </svg>
  </div>
</body>
`;
